File: src/Core/Platform/ScreenSize.c
#include <math.h>
#include "Platform.h"
#define ANSI_START           1
#define POLLING_RETRY_DELAY  10
#define EVENT_BUFFER_SIZE    256
#define RESPONSE_BUFFER_SIZE 1024

Point Platform_PixelToCell(const Platform *that, const Point pixel)
{
    return (Point) { //1 based ANSI offset
        .X = (short) (
            floorf(((float) pixel.X) * that->ScreenSize.CellsPerPixel.X) + ANSI_START
        ),
        .Y = (short) (
            floorf(((float) pixel.Y) * that->ScreenSize.CellsPerPixel.Y) + ANSI_START
        )
    };
}

Point Platform_CellToTopLeftPixel(const Platform *that, const Point cell)
{
    return (Point) { //-1 to get width upto end of previous cell
        .X = (short) (
            (float) (cell.X - ANSI_START) * that->ScreenSize.PixelsPerCell.X
        ),
        .Y = (short) (
            (float) (cell.Y - ANSI_START) * that->ScreenSize.PixelsPerCell.Y
        )
    };
}

Point Platform_FindScreenCenter(const Platform *that)
{
    return (Point) { 
        .X = that->ScreenSize.Pixels.X >> 1, 
        .Y = that->ScreenSize.Pixels.Y >> 1 
    };
}

static inline Bool Platform_HasEvents(
    const Platform *that, INPUT_RECORD *eventBuffer, Word *eventsRead)
{
    Word pendingEvents;

    return GetNumberOfConsoleInputEvents(that->In, &pendingEvents)
        && pendingEvents > 0 
        && ReadConsoleInputW(that->In, eventBuffer, EVENT_BUFFER_SIZE, eventsRead)
        && *eventsRead > 0;
}

static inline Bool Platform_IsKeyDownEvent(const INPUT_RECORD *event)
{
    return event->EventType == KEY_EVENT && event->Event.KeyEvent.bKeyDown;
}

static Bool Platform_AssignScreenSize(Platform *that, const wchar_t *respBuffer)
{
    wchar_t *pixels = wcsstr(respBuffer, L"[4;");
    wchar_t *cells = wcsstr(respBuffer, L"[8;");

    if (pixels && cells) 
    {
        swscanf(pixels, L"[4;%hd;%hdt", &that->ScreenSize.Pixels.Y, &that->ScreenSize.Pixels.X);
        swscanf(cells, L"[8;%hd;%hdt", &that->ScreenSize.Cells.Y, &that->ScreenSize.Cells.X);
        
        that->ScreenSize.CellsPerPixel = (FloatPoint) {
            .X = (float)that->ScreenSize.Cells.X / (float)that->ScreenSize.Pixels.X,
            .Y = (float)that->ScreenSize.Cells.Y / (float)that->ScreenSize.Pixels.Y
        };

        that->ScreenSize.PixelsPerCell = (FloatPoint) {
            .X = 1.0f / that->ScreenSize.CellsPerPixel.X,
            .Y = 1.0f / that->ScreenSize.CellsPerPixel.Y,
        };

        return true;
    }

    return false;
}

void Platform_UpdateScreenSize(Platform *that)
{
    FlushConsoleInputBuffer(that->In);
    Platform_DumpToConsole(that, CommandAndSize(ExitSixelMode));
    Platform_DumpToConsole(that, CommandAndSize(RequestScreenSize));
    wchar_t respBuffer[RESPONSE_BUFFER_SIZE] = {0}; int respBufferIndex = 0;
    INPUT_RECORD eventBuffer[EVENT_BUFFER_SIZE]; Word eventsRead;

    while(true)
    {
        if (Platform_HasEvents(that, eventBuffer, &eventsRead))
        {    
            for (Word i = 0; i < eventsRead; i++) 
            {  
                if (Platform_IsKeyDownEvent(&eventBuffer[i])) 
                {
                    wchar_t c = eventBuffer[i].Event.KeyEvent.uChar.UnicodeChar;
                    assert(respBufferIndex < RESPONSE_BUFFER_SIZE - 1);
                    if (c != 0) { respBuffer[respBufferIndex++] = c; }
                }
            }

            if(Platform_AssignScreenSize(that, respBuffer)) { break; }
        }
        
        Sleep(POLLING_RETRY_DELAY);
    }
}

File: CSharpPort/ClickBaitSln/ClickBait/Platform/ScreenSize.cs
using ClickBait.Drawing;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace ClickBait.Platform
{
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    internal unsafe struct ScreenSize
    {
        internal Point<short> Cells;
        internal Point<short> Pixels;
        internal Point<float> CellsPerPixel;
        internal Point<float> PixelsPerCell;
        private const ushort _ansiStart = 1;
        private const ushort _responseBufferSize = 1024;
        private const ushort _eventBufferSize = 256;
        private const ushort _pollingsRetryDelay = 10;


        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal readonly Point<short> PixelToCell(Point<short> pixel) => new()
        {
            X = (short)(MathF.Floor(pixel.X * CellsPerPixel.X) + _ansiStart),
            Y = (short)(MathF.Floor(pixel.Y * CellsPerPixel.Y) + _ansiStart)
        };

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal readonly Point<short> CellToTopLeftPixel(Point<short> cell) => new()
        {
            X = (short)((cell.X - _ansiStart) * PixelsPerCell.X),
            Y = (short)((cell.Y - _ansiStart) * PixelsPerCell.Y),
        };

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal readonly Point<short> FindScreenCenter() => new()
        {
            X = (short)(Pixels.X >> 1),
            Y = (short)(Pixels.Y >> 1)
        };

        private static bool ExtractSize(
            ReadOnlySpan<char> responseBuffer, ReadOnlySpan<char> sequence, out Point<short> size)
        {
            var index = responseBuffer.IndexOf(sequence);
            if (index != -1)
            {
                var data = responseBuffer.Slice(index + 3);
                var delimeterIndex = data.IndexOf(";");
                var endIndex = data.IndexOf("t");

                if (delimeterIndex != -1 && endIndex != -1)
                {
                    var y = data.Slice(0, delimeterIndex);
                    var x = data.Slice(delimeterIndex + 1, endIndex - delimeterIndex - 1);

                    if (short.TryParse(y, out var yValue) && short.TryParse(x, out var xValue))
                    {
                        size.Y = yValue;
                        size.X = xValue;
                        return true;
                    }
                }
            }
            size.Y = short.MinValue;
            size.X = short.MinValue;
            return false;
        }

        internal void UpdateScreenSize(in Platform platform)
        {
            FlushConsoleInputBuffer(platform.In);
            platform.DumpToConsole(Commands.ExitSixelMode());
            platform.DumpToConsole(Commands.RequestScreenSize());
            char* responseBuffer = stackalloc char[_responseBufferSize];
            int responseBufferIndex = 0;
            Unsafe.SkipInit(out uint eventsRead);
            INPUT_RECORD* eventBuffer = stackalloc INPUT_RECORD[_eventBufferSize];

            while (true)
            {
                if (platform.HasEvents(eventBuffer, &eventsRead))
                {
                    for (uint i = 0; i < eventsRead; i++)
                    {
                        if (Event.IsKeyDownEvent(in eventBuffer[i]))
                        {
                            char c = eventBuffer[i].Event.KeyEvent.uChar.UnicodeChar;
                            if (responseBufferIndex >= _responseBufferSize - 1)
                            {
                                Environment.Exit(6969);
                            }
                            if (c != '\0')
                            {
                                responseBuffer[responseBufferIndex++] = c;
                            }
                        }
                    }

                    ReadOnlySpan<char> response = new(responseBuffer, responseBufferIndex);
                    
                    if (ExtractSize(response, "[4;", out var pixels) && 
                        ExtractSize(response, "[8;", out var cells))
                    {
                        Pixels = pixels;
                        Cells = cells;
                        CellsPerPixel.X = (float)cells.X / pixels.X;
                        CellsPerPixel.Y = (float)cells.Y / pixels.Y;
                        PixelsPerCell.X = 1.0f / CellsPerPixel.X;
                        PixelsPerCell.Y = 1.0f / CellsPerPixel.Y;
                        break;
                    }
                }

                Thread.Sleep(_pollingsRetryDelay);
            }
        }
    };

}
